### PR TITLE
Add Dropbox to video providers

### PR DESCRIPTION
**Pull request by tobydriscoll**
Welcome to the quarto GitHub repo!

We are always happy to hear feedback from our users.

To file a _pull request_, please follow these instructions carefully: <https://yihui.org/issue/#bug-reports>

If you're a collaborator from outside `quarto-dev` making changes larger than a typo, please make sure you have filed an [individual](https://rstudioblog.files.wordpress.com/2017/05/rstudio_individual_contributor_agreement.pdf) or [corporate](https://rstudioblog.files.wordpress.com/2017/05/rstudio_corporate_contributor_agreement.pdf) contributor agreement. You can send the signed copy to jj@rstudio.com.

Also, please complete and keep the checklist below.

## Description

Addresses issue #4153.

To embed a video saved on dropbox.com, use Dropbox to get a shareable link to the file. Then paste that link into a video shortcode, e. g., 

```
{{< video https://www.dropbox.com/s/f7kzi3u0hvnkldr/5_16_08_19.mp4?raw=1 >}}
```
**Note:** The link as pasted from Dropbox will end in `?dl=0`, which must be replaced with `?raw=1` as in the above example.

## Checklist

I have (if applicable):

- [ x] filed a [contributor agreement](https://github.com/quarto-dev/quarto-cli/blob/main/CONTRIBUTING.md).
- [ x] referenced the GitHub issue this PR closes
- [ ] updated the appropriate changelog

### CODE REVIEW

**Comment by cscheid** [comment]
Thanks for the PR! Would you consider detecting the `?dl=0` URL and rewriting it with `?raw=1` in your code?

**Comment by tobydriscoll** [comment]
It's easy to see how to gsub the raw=1 string into place, but I could not figure out how to assign that result to something in a way that replaces the input `params.src` field. Definitely not a Lua coder!

**Comment by cscheid** [comment]
That's ok. I didn't want to step on your code before asking but I'll give it a shot.

**Comment by cscheid** [comment]
Ok, I've given this a quick review, and I don't think this PR works as is.

It seems that for me, on the URL you shared, the videoJS handler is being called before dropboxBuilder, and when `?raw=1` is set, the videoJS handler is enough. But if we switch the order so that your new helper function is called, then the resulting video link doesn't appear to work as well as videoJS; I get a download prompt on the VS code preview, and a badly proportioned video in Chrome.

I've also confirmed that the `?raw=1` link works on `main` without the PR:

image

So, I'm a bit confused.

File: src/resources/extensions/quarto/video/video.lua [hunk 1]
```diff
@@ -152,12 +153,32 @@ local videoJSBuilder = function(params)
...
+  local src = params.src
+  local isDropbox = function()
+    return string.find(src, 'https://www.dropbox.com') and checkEndMatch(src, '?raw=1')
+  end
+
+  if not isDropbox() then return nil end
+
+  local result = {}
+
+  local SNIPPET = [[<iframe data-external="1" src="{src}"{width}{height} allowfullscreen="" title="{title}" allow="encrypted-media"></iframe>]]
```
Comment (allenmanning) [inline src/resources/extensions/quarto/video/video.lua:168]: Please consider adding this to the unit tests.

Comment (tobydriscoll) [reply src/resources/extensions/quarto/video/video.lua:168]: I'm confused too. 

I've just tried again pasting my video code into the `index.qmd` of a newly created book project, and `quarto preview` in version 1.2.280 still gives me `No video source specified for video shortcode` and refuses to continue the preview. The version built from my PR works fine. 

I don't know anything about Lua or videoJS, at all, so I'm at a loss.

Comment (cscheid) [reply src/resources/extensions/quarto/video/video.lua:168]: @tobydriscoll You need to test against 1.3 prereleases.

**Comment by tobydriscoll** [comment]
I guess I have to tap out. The contributing guidelines made me think I was supposed to branch from main to add something. Anyway, it sounds like v1.3 won't have the issue I was trying to fix, so it's probably moot. I'm happy to limp along with my private frankenstein version until then.

**Comment by cscheid** [comment]
> I guess I have to tap out. The contributing guidelines made me think I was supposed to branch from main to add something.

That's actually the case: the 1.3 prereleases are built from main. It's 1.2 that's on the `v1.2` branch. If that part of the documentation confused you, we should probably update it. Can you point me to where you saw this? Thank you!

**Pull request closed without merging**